src/internal/pfsdb/common.go

```diff
@@ -0,0 +1,199 @@
+	Equal              FilterOperator = "%s = ?"
+	GreaterThan        FilterOperator = "%s > ?"
+	GreaterThanOrEqual FilterOperator = "%s >= ?"
+	LessThan           FilterOperator = "%s < ?"
+	LessThanOrEqual    FilterOperator = "%s <= ?"
+	NotEqual           FilterOperator = "%s <> ?"
+	ValueIn            FilterOperator = "%s in (?)"
+)
+
+type (
+	ModelType interface{ Repo | Commit | Branch }
+	FieldType interface{ CommitField | BranchField }
+	PairType  interface{ CommitPair | BranchPair }
+)
+
+type Filter[T FieldType] struct {
+	Field  T
+	Value  any
+	values []any
+	Op     FilterOperator
+}
+
+func (f *Filter[T]) QueryString() (formatted string, err error) {
+	formatted = fmt.Sprintf(string(f.Op), f.Field)
+	if f.Op == ValueIn {
+		formatted, f.values, err = sqlx.In(formatted, f.Value) // Thank god for this!
+		if err != nil {
+			return "", errors.Wrapf(err, "expanding %s", formatted)
+		}
+	}
+	return
+}
+
+type AndFilters[T FieldType] []*Filter[T]
+
+func (f AndFilters[T]) QueryString() (string, error) {
+	var parts []string
+	for _, filter := range f {
+		part, err := filter.QueryString()
+		if err != nil {
+			return "", err
+		}
+		parts = append(parts, part)
+	}
+	return strings.Join(parts, " AND "), nil
+}
+
+type OrderBy[T FieldType] struct {
+	Fields    []T
+	SortOrder SortOrder
+}
+
+func (o *OrderBy[T]) QueryString() string {
+	var parts []string
+	for _, field := range o.Fields {
+		parts = append(parts, string(field))
+	}
+	var order string
+	switch o.SortOrder {
+	case SortAscend:
+		order = "ASC"
+	case SortDescend:
+		order = "DESC"
+	}
+	return strings.Join(parts, ", ") + " " + order
+}
+
+type QueryBuilder[T FieldType] struct {
+	baseQuery   string
+	queryParams []any
+
+	AndFilters AndFilters[T]
+	GroupBy    string
+	OrderBy    *OrderBy[T]
+	// TODO, should QueryBuilder deal with Limit and Offset?
+}
+
+func (c *QueryBuilder[T]) Build() (string, error) {
+	query := c.baseQuery
+	condition, err := c.AndFilters.QueryString()
+	if err != nil {
+		return "", err
+	}
+	if condition != "" {
+		for _, filter := range c.AndFilters {
+			if filter.Op == ValueIn {
+				// filter.values is set by Filter.String()
+				c.queryParams = append(c.queryParams, filter.values...)
+			} else {
+				c.queryParams = append(c.queryParams, filter.Value)
+			}
+		}
+		query += "\nWHERE " + condition
+	}
+	// TODO implement GROUP BY
+	if c.GroupBy != "" {
+		query += "\nGROUP BY " + c.GroupBy
+	}
+	if c.OrderBy != nil {
+		query += "\nORDER BY " + c.OrderBy.QueryString()
+	}
+	query = sqlx.Rebind(sqlx.DOLLAR, query)
+	return query, nil
+}
+
+type transformFn[T ModelType, U PairType] func(context.Context, *pachsql.Tx, *T) (*U, error)
+
+// T is the type used to deserialize rows from the database
+// U is the type the type returned by the iterator
+// Typically the user supplies the transform function, which converts T to U
+type pageIterator[T ModelType, U PairType] struct {
```

Comment:
I don't think this is the right abstraction over pages.  There shouldn't need to be a helper object that is used to implement the other iterators; this should do everything.  `Iterator[T]` should be able to get any kind of `T` with only a query (that limit/offset is safe to append to).  You also don't need to cache the entire page; keep the Rows object and StructScan into the destination at yield time.  If rows.Next() is false, then do the next page.  If that's ErrNoRows, then EOS.

A `Close` method is missing from the stream.Iterator API, for giving up early (context.Canceled, etc.).  That needs to be added and needs to close the Rows object.

Reply:
So the reason we need to cache the page is to avoid holding a connection mid-page. We share a single tx object across all pages, for each element of a page, we might want to run some extra db calls to obtain additional info. If we hold on to the `rows` object (which holds on to a connection), then any subsequent use of `tx` for other queries would return a "conn is busy error".